### PR TITLE
scheduler: establish node topology info cache

### PR DESCRIPTION
**Pull request by shiyan2016**
**What type of PR is this?**
 /kind design
 /sig scheduling

**What this PR does / why we need it**:

This PR tries to establish a cache mapping topology pairs to nodes used by https://github.com/kubernetes/kubernetes/pull/79474.

**Special notes for your reviewer**:
This PR is a pre-PR for https://github.com/kubernetes/kubernetes/pull/79474.

### CODE REVIEW

**Review by draveness — COMMENTED** [review]
/kind feature
/ok-to-test
/release-note-none

File: pkg/scheduler/internal/cache/node_topology_info.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+func (t NodeTopologyInfo) addTopologyPair(pair TopologyPair, nodeName string) {
+	if _, ok := t[pair]; !ok {
+		t[pair] = sets.String{}
+	}
+	t[pair][nodeName] = sets.Empty{}
+}
+
+func (t NodeTopologyInfo) removeTopologyPair(pair TopologyPair, nodeName string) {
+	if nodeNameSet, ok := t[pair]; ok {
+		delete(nodeNameSet, nodeName)
```
Comment (tedyu) [inline pkg/scheduler/internal/cache/node_topology_info.go]: Maybe the func name can be more descriptive: in normal case nodeName is removed, not the pair

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -1247,7 +1250,18 @@ func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod,
 		if len(term.TopologyKey) == 0 {
 			return false, false, fmt.Errorf("empty topologyKey is not allowed except for PreferredDuringScheduling pod anti-affinity")
 		}
-		if !priorityutil.NodesHaveSameTopologyKey(nodeInfo.Node(), targetPodNode, term.TopologyKey) {
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/predicates/predicates.go]: In theory the operation is O(n), but in practice a node usually have ~10 label pairs, so I'm not quite a solution like this PR with O(1) implementation helps a lot. Your test result seems a little strange, I will run the same test again.

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: @shiyan2016 testing in my local 32G-8cpu BM shows a different result:

```
## Without this PR

BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         131,230,945 ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         133,996,504 ns/op

BenchmarkScheduling/5000Nodes/1000Pods-8                  1000          89,063,795 ns/op
BenchmarkScheduling/5000Nodes/1000Pods-8                  1000          91,836,182 ns/op

## With this PR

BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         128,431,629 ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         134,703,083 ns/op

BenchmarkScheduling/5000Nodes/1000Pods-8                  1000          93,030,224 ns/op
BenchmarkScheduling/5000Nodes/1000Pods-8                  1000          88,811,118 ns/op
```

Comment (draveness) [reply pkg/scheduler/algorithm/predicates/predicates.go]: testing in my local 16G-4CPU shows there is not that much performance difference with or without this PR.

## Without this PR

```go
BenchmarkScheduling/5000Nodes/1000Pods-8                             1000	  44703105 ns/op
BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8    1000	  48068337 ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8           1000	  94545001 ns/op
BenchmarkSchedulingNodeAffinity/5000Nodes/1000Pods-8        1000	  45828025 ns/op
```

## With this PR

```go
BenchmarkScheduling/5000Nodes/1000Pods-8                           1000	  46596568 ns/op
BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8  1000	  46290482 ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8         1000	  95529805 ns/op
BenchmarkSchedulingNodeAffinity/5000Nodes/1000Pods-8       1000	  45912117 ns/op
```

**Comment by shiyan2016** [comment]
@Huang-Wei @draveness I did the benchmark test on the old vm and two new machines. The results on the old vm(2c6g) still has 2X perfermance improvement. But on the other two new machine(4c8g, 8c8g). The results are as follows:
```
1. 2c6g vm
- without this pr
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-2         	    1000	 519818840 ns/op
- with this pr
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-2         	    1000	 239119822 ns/op
2. 4c8g bm
- without this pr
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-4                 1000	  197792912 ns/op
- with this pr
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-4                 1000	  156623295 ns/op
3. 8c8g vm
- without this pr
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         168338872 ns/op
- with this pr
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         158413791 ns/op
```
From the results, it shows that this PR does not helps a lot for performance improvement. But this PR is still useful because https://github.com/kubernetes/kubernetes/pull/79474 will use the nodecache to convert topology pairs scores to the InterPodAffinityPriority node scores. 

To avoid submitting one big PR,  I still reserve this one.  I changed the title and removed the replacing of func NodesHaveSameTopologyKey. Please review it continuely. Thanks.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 2 above)
Comment (shiyan2016) [reply pkg/scheduler/algorithm/predicates/predicates.go]: > In theory the operation is O(n), but in practice a node usually have ~10 label pairs, so I'm not quite a solution like this PR with O(1) implementation helps a lot. Your test result seems a little strange, I will run the same test again.

Replace NodesHaveSameTopologyKey using nodecache is mainly used to reduce node iteration and topology pair match check in InterPodAffinity processTerm: https://github.com/kubernetes/kubernetes/pull/79360/files#diff-ac123860716bbbb4ed0f26cc27062bb7R99.

Comment (draveness) [reply pkg/scheduler/algorithm/predicates/predicates.go]: > > In theory the operation is O(n), but in practice a node usually have ~10 label pairs, so I'm not quite a solution like this PR with O(1) implementation helps a lot. Your test result seems a little strange, I will run the same test again.
> 
> Replace NodesHaveSameTopologyKey using nodecache is mainly used to reduce node iteration and topology pair match check in InterPodAffinity processTerm: https://github.com/kubernetes/kubernetes/pull/79360/files#diff-ac123860716bbbb4ed0f26cc27062bb7R99.

I'm quite curious about how did you run the benchmark with 2x performance improvement. Why did we get a different result on different machines?

**Comment by draveness** [comment]
> @Huang-Wei @draveness I did the benchmark test on the old vm and two new machines. The results on the old vm(2c6g) still has 2X perfermance improvement. But on the other two new machine(4c8g, 8c8g). The results are as follows:
> 
> ```
> 1. 2c6g vm
> - without this pr
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-2         	    1000	 519818840 ns/op
> - with this pr
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-2         	    1000	 239119822 ns/op
> 2. 4c8g bm
> - without this pr
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-4                 1000	  197792912 ns/op
> - with this pr
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-4                 1000	  156623295 ns/op
> 3. 8c8g vm
> - without this pr
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         168338872 ns/op
> - with this pr
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         158413791 ns/op
> ```
> 
> From the results, it shows that this PR does not helps a lot for performance improvement. But this PR is still useful because #79474 will use the nodecache to convert topology pairs scores to the InterPodAffinityPriority node scores.
> 
> To avoid submitting one big PR, I still reserve this one. I will change the title and remove the replacing of func NodesHaveSameTopologyKey.

IMO it's better to merge #79474 into this one and run a benchmark against the merged PR since this one could not improve performance by itself.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 2 above)
Comment (shiyan2016) [reply pkg/scheduler/algorithm/predicates/predicates.go]: I also confused for the results on 2c6g. Maybe the cpu cores is the factor.

**Comment by shiyan2016** [comment]
> > @Huang-Wei @draveness I did the benchmark test on the old vm and two new machines. The results on the old vm(2c6g) still has 2X perfermance improvement. But on the other two new machine(4c8g, 8c8g). The results are as follows:
> > ```
> > 1. 2c6g vm
> > - without this pr
> > BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-2         	    1000	 519818840 ns/op
> > - with this pr
> > BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-2         	    1000	 239119822 ns/op
> > 2. 4c8g bm
> > - without this pr
> > BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-4                 1000	  197792912 ns/op
> > - with this pr
> > BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-4                 1000	  156623295 ns/op
> > 3. 8c8g vm
> > - without this pr
> > BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         168338872 ns/op
> > - with this pr
> > BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         158413791 ns/op
> > ```
> > 
> > 
> > From the results, it shows that this PR does not helps a lot for performance improvement. But this PR is still useful because #79474 will use the nodecache to convert topology pairs scores to the InterPodAffinityPriority node scores.
> > To avoid submitting one big PR, I still reserve this one. I will change the title and remove the replacing of func NodesHaveSameTopologyKey.
> 
> IMO it's better to merge #79474 into this one and run a benchmark against the merged PR since this one could not improve performance by itself.

The benchmark for both nodecache and abandon iteration result: https://github.com/kubernetes/kubernetes/pull/79474#issue-292429674.

**Comment by snowplayfire** [comment]
@draveness   #79474 has merged this PR into one,  you can come to #79474 and review it.

**Comment by draveness** [comment]
> @draveness #79474 has merged this PR into one, you can come to #79474 and review it.

Sure, I'll run the benchmark against #79474, and is it ok to close this PR?

**Comment by shiyan2016** [comment]
> > @draveness #79474 has merged this PR into one, you can come to #79474 and review it.
> 
> Sure, I'll run the benchmark against #79474 but please ensure the tests are passed, and is it ok to close this PR?

ok, I closed it.

**Pull request closed without merging**